File: app/src/main/java/com/tunjid/fingergestures/models/TextLink.java
package com.tunjid.fingergestures.models;

import androidx.annotation.NonNull;

public class TextLink implements CharSequence {

    private final CharSequence text;
    public final String link;

    public TextLink(CharSequence text, String link) {
        this.text = text;
        this.link = link;
    }

    @Override
    @NonNull
    public String toString() {
        return text.toString();
    }

    @Override
    public int length() {
        return text.length();
    }

    @Override
    public char charAt(int index) {
        return text.charAt(index);
    }

    @Override
    public CharSequence subSequence(int start, int end) {
        return text.subSequence(start, end);
    }
}

File: app/src/main/java/com/tunjid/fingergestures/models/TextLink.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.models

import com.tunjid.androidx.recyclerview.diff.Differentiable

data class TextLink(
        private val text: CharSequence,
        val link: String
) : CharSequence, Differentiable {

    override val diffId: String get() = link

    override fun areContentsTheSame(other: Differentiable): Boolean =
            (other as? TextLink)?.let{it == this} ?: super.areContentsTheSame(other)

    override fun toString(): String = text.toString()

    override val length: Int get() = text.length

    override fun get(index: Int): Char = text[index]

    override fun subSequence(startIndex: Int, endIndex: Int): CharSequence = text.subSequence(startIndex, endIndex)
}
